File: backend/testing/http_fingerprint.py
"""HTTP request/response fingerprint emitter for scripted scenarios.

For each endpoint a scenario exercises, the fingerprinter records a
canonical (request, normalised-response) pair and asserts it against
a per-endpoint golden file under
``<scenario>/expected/http_responses/<endpoint_id>.json``. The
``--update-fingerprints`` pytest flag (already registered by the
backend-root conftest) flips the workflow into write mode for
deliberate golden ratification.

The fingerprint reuses the shared ``Normalizer`` so a UUID or
timestamp the bus surfaced first resolves to the same symbol when it later
appears in an HTTP response body; that keeps the HTTP goldens
cross-referenceable with the per-scenario fingerprint.jsonl + db_state.json.

The response header projection is intentionally narrow: only the
substrate's own headers (``Content-Type``, ``Cache-Control``,
``ETag``) are pinned. ETag is recorded as a presence-and-shape
sentinel (``"<STRONG_ETAG>"``) rather than its literal hex, because
the hex hashes a body that contains UUIDs and timestamps which
change per run; the body's own normalisation pins identity, and the
header pins the substrate is engaged.

Binary bodies (e.g. ``image/png`` from ``/api/scan/skills/capture/{page}``)
are projected as ``{"_binary": true, "byte_length": N}``; pinning byte
length lets the goldens catch a content-shape regression without
embedding the raw bytes.
"""

from __future__ import annotations

import json
import re
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

from backend.testing.fingerprint import Normalizer

STRONG_ETAG_RE = re.compile(r'^"[0-9a-f]{64}"$')

# Inline UUID pattern (anchored to a path segment); the ``UUID_PATTERN``
# imported above is the full-string match used by the body walker.
_UUID_IN_PATH_RE = re.compile(
    r"[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}"
)

# The header projection: only these are written into the golden.
# Anything else (Server, Date, CORS) varies across environments and
# would make the goldens noisy without strengthening the contract.
PROJECTED_HEADERS: tuple[str, ...] = (
    "content-type",
    "cache-control",
    "etag",
)

ETAG_SENTINEL = "<STRONG_ETAG>"

# A tracking session's ``duration`` is ``ended_at - started_at`` in whole
# seconds, both captured from wall-clock at session start / stop, so its value
# reflects how long the test happened to run (its replay-drain latency) rather
# than any contract the response should pin. The sibling ``startTime`` /
# ``endTime`` are already symbolised by the shared Normalizer; this projects
# the derived delta the same way so the golden stays deterministic regardless
# of how quickly the watcher drained the scenario.
SESSION_DURATION_SENTINEL = "<SESSION_DURATION>"


@dataclass(frozen=True)
class HttpRequest:
    """The request side of a captured (request, response) pair."""

    method: str
    path: str
    query: dict[str, Any] = field(default_factory=dict)
    headers: dict[str, str] = field(default_factory=dict)


@dataclass(frozen=True)
class HttpResponse:
    """The response side of a captured pair, post-normalisation."""

    status_code: int
    headers: dict[str, str]
    body: Any


class HttpFingerprintMismatch(AssertionError):
    """Raised when a captured fingerprint diverges from its golden.

    Carries the endpoint_id and a diff-ready expected/actual pair so a
    downstream renderer can surface the divergence without re-parsing
    the formatted message.
    """

    def __init__(self, endpoint_id: str, expected: Any, actual: Any) -> None:
        self.endpoint_id = endpoint_id
        self.expected = expected
        self.actual = actual
        expected_text = json.dumps(
            expected, indent=2, sort_keys=True, ensure_ascii=False
        )
        actual_text = json.dumps(actual, indent=2, sort_keys=True, ensure_ascii=False)
        super().__init__(
            f"HTTP fingerprint mismatch for endpoint {endpoint_id!r}.\n\n"
            f"--- expected (golden) ---\n{expected_text}\n\n"
            f"--- actual (this run) ---\n{actual_text}\n\n"
            "Rerun with `pytest --update-fingerprints` (and review the diff) "
            "if the new output is the intended new golden."
        )


def normalise_etag(value: str | None) -> str | None:
    """Project an ETag header into the golden's canonical form.

    A strong-format ETag (``"<sha256-hex>"``) is replaced by the
    sentinel ``"<STRONG_ETAG>"``; that pins the substrate is engaged
    without binding the golden to the hex of a body that contains
    per-run UUIDs and timestamps. Any other shape (a weak ETag, an
    empty string, or a malformed value) is kept verbatim so an
    unexpected shape surfaces as a divergence rather than silently
    coercing into the sentinel.
    """
    if value is None:
        return None
    if STRONG_ETAG_RE.match(value):
        return ETAG_SENTINEL
    return value


def project_headers(raw_headers: dict[str, str]) -> dict[str, str]:
    """Filter ``raw_headers`` down to the projection the golden tracks."""
    lowered = {k.lower(): v for k, v in raw_headers.items()}
    projected: dict[str, str] = {}
    for name in PROJECTED_HEADERS:
        if name not in lowered:
            continue
        value = lowered[name]
        if name == "etag":
            etag_projection = normalise_etag(value)
            if etag_projection is not None:
                projected[name] = etag_projection
        else:
            projected[name] = value
    return projected


def normalise_path(path: str, normalizer: Normalizer) -> str:
    """Replace each UUID segment in ``path`` with its ``<UUID_N>`` symbol.

    Path-side normalisation must share the body's symbol table so the
    same session UUID surfacing in ``/api/tracking/session/<id>`` and
    in the response body's ``"sessionId"`` resolves to the same symbol.
    """

    def _replace(match: re.Match[str]) -> str:
        result = normalizer.normalize(match.group(0))
        return result if isinstance(result, str) else match.group(0)

    return _UUID_IN_PATH_RE.sub(_replace, path)


def normalise_body(
    raw_body: bytes,
    content_type: str | None,
    normalizer: Normalizer,
) -> Any:
    """Render the response body into the golden's canonical form.

    JSON bodies are parsed and walked through the shared
    ``Normalizer`` so UUIDs / timestamps land on the same symbols
    the per-scenario fingerprint already uses. Empty bodies are emitted as
    ``null``. Anything else is treated as binary and projected as
    ``{"_binary": true, "byte_length": N}`` so the byte-shape is
    pinned without embedding the raw bytes (which would defeat the
    UUID / timestamp normalisation rationale for everything else).
    """
    if not raw_body:
        return None
    if content_type and "application/json" in content_type.lower():
        decoded = json.loads(raw_body.decode("utf-8"))
        return _project_session_duration(normalizer.normalize(decoded))
    return {"_binary": True, "byte_length": len(raw_body)}


def _project_session_duration(value: Any) -> Any:
    """Symbolise the wall-clock session ``duration`` to a stable sentinel.

    Walks the normalised body and replaces any numeric ``duration`` with
    ``SESSION_DURATION_SENTINEL``. Across the hydration surface the goldens
    capture, ``duration`` is only ever a tracking session's wall-clock length
    (in the sessions list and in session detail's summary), so a value-level
    pin would encode replay-drain latency rather than a contract; the sentinel
    keeps the presence-and-shape assertion without the non-determinism. A
    future endpoint that emits a genuinely deterministic ``duration`` would
    surface the sentinel here and can refine this projection then.
    """
    if isinstance(value, dict):
        projected = {k: _project_session_duration(v) for k, v in value.items()}
        if isinstance(projected.get("duration"), int | float) and not isinstance(
            projected.get("duration"), bool
        ):
            projected["duration"] = SESSION_DURATION_SENTINEL
        return projected
    if isinstance(value, list):
        return [_project_session_duration(item) for item in value]
    return value


@dataclass(frozen=True)
class HttpCapture:
    """One (request, normalised-response) pair recorded for a scenario."""

    request: HttpRequest
    response: HttpResponse

    def to_golden_dict(self) -> dict[str, Any]:
        """Render this capture as the dict the golden file holds.

        Keys are sorted in the serialised JSON, so two captures of
        the same response produce byte-identical golden text.
        """
        return {
            "request": {
                "method": self.request.method,
                "path": self.request.path,
                "query": dict(sorted(self.request.query.items())),
            },
            "response": {
                "status_code": self.response.status_code,
                "headers": dict(sorted(self.response.headers.items())),
                "body": self.response.body,
            },
        }


class HttpFingerprinter:
    """Capture HTTP requests against per-endpoint goldens for a scenario.

    The fingerprinter is bound to one ``scenario_dir`` for its lifetime;
    each ``capture`` call writes to (or compares against)
    ``<scenario_dir>/expected/http_responses/<endpoint_id>.json``. The
    normaliser is shared with the caller's existing fingerprint
    substrate so the symbol table is consistent across the per-scenario
    event-stream fingerprint, DB snapshot, and HTTP captures.
    """

    def __init__(
        self,
        scenario_dir: Path,
        normalizer: Normalizer,
        *,
        update: bool = False,
    ) -> None:
        self.scenario_dir = scenario_dir
        self.expected_dir = scenario_dir / "expected" / "http_responses"
        self.normalizer = normalizer
        self._update = update
        self._captured: list[str] = []

    @property
    def update_mode(self) -> bool:
        """True when the run was invoked with ``--update-fingerprints``."""
        return self._update

    @property
    def captured_endpoint_ids(self) -> tuple[str, ...]:
        """The endpoint_ids captured so far this run (in invocation order)."""
        return tuple(self._captured)

    def capture(
        self,
        response: Any,
        *,
        endpoint_id: str,
        request_method: str,
        request_path: str,
        request_query: dict[str, Any] | None = None,
    ) -> HttpCapture:
        """Normalise ``response`` and assert it against the endpoint's golden.

        ``response`` is a ``httpx.Response`` (TestClient yields these);
        the fingerprinter reads ``status_code``, ``headers``, and
        ``content`` so any compatible response object works. Returns
        the captured ``HttpCapture`` so callers can layer extra
        assertions on top (status range, body shape) without re-doing
        the normalisation.
        """
        content_type = response.headers.get("content-type")
        headers = project_headers(dict(response.headers))
        body = normalise_body(response.content, content_type, self.normalizer)
        normalised_path = normalise_path(request_path, self.normalizer)

        capture = HttpCapture(
            request=HttpRequest(
                method=request_method,
                path=normalised_path,
                query=dict(request_query or {}),
            ),
            response=HttpResponse(
                status_code=response.status_code,
                headers=headers,
                body=body,
            ),
        )

        self._captured.append(endpoint_id)
        golden_path = self.expected_dir / f"{endpoint_id}.json"
        actual = capture.to_golden_dict()

        if self._update:
            self._write_golden(golden_path, actual)
            return capture

        if not golden_path.exists():
            raise HttpFingerprintMismatch(
                endpoint_id=endpoint_id,
                expected=None,
                actual=actual,
            )

        expected = json.loads(golden_path.read_text(encoding="utf-8"))
        if expected != actual:
            raise HttpFingerprintMismatch(
                endpoint_id=endpoint_id,
                expected=expected,
                actual=actual,
            )
        return capture

    def _write_golden(self, path: Path, payload: dict[str, Any]) -> None:
        """Serialise ``payload`` to ``path`` as canonical sorted JSON."""
        path.parent.mkdir(parents=True, exist_ok=True)
        text = json.dumps(payload, sort_keys=True, indent=2, ensure_ascii=False) + "\n"
        path.write_text(text, encoding="utf-8")

File: frontend/src-tauri/eo-wire/src/http_fingerprint.rs
//! HTTP request/response fingerprint emitter: Rust port of
//! `backend/testing/http_fingerprint.py`.
//!
//! For one captured response this reproduces the canonical
//! `(request, normalised-response)` golden the committed
//! `expected/http_responses/<endpoint_id>.json` holds: the header projection
//! (`Content-Type`/`Cache-Control`/`ETag`, with a strong ETag reduced to the
//! `<STRONG_ETAG>` sentinel), the UUID-segment path normalisation, the body
//! walked through the shared [`Normalizer`] (with the wall-clock session
//! `duration` reduced to `<SESSION_DURATION>`), and binary bodies projected to
//! their byte length.

use serde_json::{Map, Value};

use crate::normalizer::{to_python_json, Normalizer};

/// Only these response headers are pinned (lower-cased); everything else
/// varies across environments. Mirrors `PROJECTED_HEADERS`.
const PROJECTED_HEADERS: [&str; 3] = ["content-type", "cache-control", "etag"];

const ETAG_SENTINEL: &str = "<STRONG_ETAG>";
const SESSION_DURATION_SENTINEL: &str = "<SESSION_DURATION>";

/// One captured response in the wire-portable form the raw-capture fixture
/// holds: the request line plus the raw response (status, headers, body bytes).
pub struct RawResponse<'a> {
    pub method: &'a str,
    pub path: &'a str,
    pub query: &'a Map<String, Value>,
    pub status_code: i64,
    pub headers: &'a Map<String, Value>,
    pub body: &'a [u8],
}

/// Whether `value` is a strong-format ETag (`"<sha256-hex>"`), matching
/// `STRONG_ETAG_RE` (`^"[0-9a-f]{64}"$`).
fn is_strong_etag(value: &str) -> bool {
    let bytes = value.as_bytes();
    if bytes.len() != 66 {
        return false;
    }
    if bytes[0] != b'"' || bytes[65] != b'"' {
        return false;
    }
    bytes[1..65]
        .iter()
        .all(|&c| c.is_ascii_digit() || (b'a'..=b'f').contains(&c))
}

/// Project an ETag header into the golden's canonical form (strong -> sentinel,
/// anything else verbatim). Mirrors `normalise_etag`.
fn normalise_etag(value: &str) -> String {
    if is_strong_etag(value) {
        ETAG_SENTINEL.to_string()
    } else {
        value.to_string()
    }
}

/// Filter raw headers to the projection the golden tracks, in `PROJECTED_HEADERS`
/// order. Header names are matched case-insensitively; a missing one is skipped.
/// Mirrors `project_headers` (which builds a dict later serialised sorted).
fn project_headers(raw_headers: &Map<String, Value>) -> Map<String, Value> {
    let mut lowered: Map<String, Value> = Map::new();
    for (key, value) in raw_headers {
        lowered.insert(key.to_lowercase(), value.clone());
    }
    let mut projected = Map::new();
    for name in PROJECTED_HEADERS {
        let Some(Value::String(value)) = lowered.get(name) else {
            continue;
        };
        if name == "etag" {
            projected.insert(name.to_string(), Value::String(normalise_etag(value)));
        } else {
            projected.insert(name.to_string(), Value::String(value.clone()));
        }
    }
    projected
}

/// Whether the 36 bytes at `pos` form a canonical lowercase-hex UUID (the
/// inline, unanchored `_UUID_IN_PATH_RE` shape).
fn uuid_at(bytes: &[u8], pos: usize) -> bool {
    if pos + 36 > bytes.len() {
        return false;
    }
    for offset in 0..36 {
        let c = bytes[pos + offset];
        let ok = match offset {
            8 | 13 | 18 | 23 => c == b'-',
            _ => c.is_ascii_digit() || (b'a'..=b'f').contains(&c),
        };
        if !ok {
            return false;
        }
    }
    true
}

/// Replace each UUID segment in `path` with its `<UUID_N>` symbol, sharing the
/// body's symbol table. Mirrors `normalise_path`'s `re.sub` over the inline
/// (unanchored, fixed-length) UUID pattern: leftmost non-overlapping matches.
fn normalise_path(path: &str, normalizer: &mut Normalizer) -> String {
    let bytes = path.as_bytes();
    let mut out = String::with_capacity(path.len());
    let mut i = 0;
    while i < bytes.len() {
        if uuid_at(bytes, i) {
            let uuid = &path[i..i + 36];
            let symbol = normalizer.normalize(&Value::String(uuid.to_string()));
            match symbol {
                Value::String(s) => out.push_str(&s),
                _ => out.push_str(uuid),
            }
            i += 36;
        } else {
            // ASCII path; UUID runs are ASCII so a byte step never splits a
            // multi-byte char at a match site. Copy this byte's char.
            let ch = path[i..].chars().next().expect("non-empty remainder");
            out.push(ch);
            i += ch.len_utf8();
        }
    }
    out
}

/// Render the response body into the golden's canonical form. JSON bodies are
/// parsed, normalised, and have their session `duration` projected; empty
/// bodies become `null`; anything else projects to `{"_binary": true,
/// "byte_length": N}`. Mirrors `normalise_body`.
fn normalise_body(body: &[u8], content_type: Option<&str>, normalizer: &mut Normalizer) -> Value {
    if body.is_empty() {
        return Value::Null;
    }
    if let Some(ct) = content_type {
        if ct.to_lowercase().contains("application/json") {
            let decoded: Value = serde_json::from_slice(body).expect("json body parses as JSON");
            return project_session_duration(normalizer.normalize(&decoded));
        }
    }
    let mut binary = Map::new();
    binary.insert("_binary".to_string(), Value::Bool(true));
    binary.insert("byte_length".to_string(), Value::Number(body.len().into()));
    Value::Object(binary)
}

/// Replace any numeric `duration` with the wall-clock sentinel. Mirrors
/// `_project_session_duration` (post-normalisation, so an epoch `duration`
/// already symbolised to a string is left alone; only a numeric one is hit).
fn project_session_duration(value: Value) -> Value {
    match value {
        Value::Object(map) => {
            let mut out = Map::new();
            for (key, val) in map {
                out.insert(key, project_session_duration(val));
            }
            if out.get("duration").map(Value::is_number).unwrap_or(false) {
                out.insert(
                    "duration".to_string(),
                    Value::String(SESSION_DURATION_SENTINEL.to_string()),
                );
            }
            Value::Object(out)
        }
        Value::Array(items) => {
            Value::Array(items.into_iter().map(project_session_duration).collect())
        }
        other => other,
    }
}

/// Normalise a captured response into its golden value. The body is normalised
/// before the path, matching the Python `capture`'s order (so a UUID first seen
/// in the body owns the lower symbol when it recurs in the path).
pub fn capture(raw: &RawResponse, normalizer: &mut Normalizer) -> Value {
    let content_type = raw
        .headers
        .iter()
        .find(|(k, _)| k.to_lowercase() == "content-type")
        .and_then(|(_, v)| v.as_str());

    let headers = project_headers(raw.headers);
    let body = normalise_body(raw.body, content_type, normalizer);
    let path = normalise_path(raw.path, normalizer);

    let mut request = Map::new();
    request.insert("method".to_string(), Value::String(raw.method.to_string()));
    request.insert("path".to_string(), Value::String(path));
    request.insert("query".to_string(), Value::Object(raw.query.clone()));

    let mut response = Map::new();
    response.insert(
        "status_code".to_string(),
        Value::Number(raw.status_code.into()),
    );
    response.insert("headers".to_string(), Value::Object(headers));
    response.insert("body".to_string(), body);

    let mut golden = Map::new();
    golden.insert("request".to_string(), Value::Object(request));
    golden.insert("response".to_string(), Value::Object(response));
    Value::Object(golden)
}

/// Serialise a captured golden value as the committed golden text (sorted keys,
/// 2-space indent, trailing newline), matching `HttpFingerprinter._write_golden`.
pub fn serialize_capture(capture: &Value) -> String {
    to_python_json(capture, Some(2)) + "\n"
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    fn headers(pairs: &[(&str, &str)]) -> Map<String, Value> {
        pairs
            .iter()
            .map(|(k, v)| (k.to_string(), Value::String(v.to_string())))
            .collect()
    }

    #[test]
    fn strong_etag_reduces_to_sentinel() {
        let hex = "a".repeat(64);
        assert!(is_strong_etag(&format!("\"{hex}\"")));
        assert_eq!(normalise_etag(&format!("\"{hex}\"")), "<STRONG_ETAG>");
        // A weak ETag is kept verbatim so an unexpected shape surfaces.
        assert_eq!(normalise_etag("W/\"abc\""), "W/\"abc\"");
    }

    #[test]
    fn headers_project_to_the_three_pinned_lowercased() {
        let raw = headers(&[
            ("Content-Type", "application/json"),
            ("Cache-Control", "no-cache"),
            ("Content-Length", "42"),
            ("Server", "uvicorn"),
        ]);
        let projected = project_headers(&raw);
        assert_eq!(projected.len(), 2);
        assert_eq!(projected["content-type"], json!("application/json"));
        assert_eq!(projected["cache-control"], json!("no-cache"));
        assert!(!projected.contains_key("content-length"));
    }

    #[test]
    fn path_uuid_segment_becomes_symbol() {
        let mut norm = Normalizer::new();
        let path = normalise_path(
            "/api/tracking/session/11111111-1111-1111-1111-111111111111/quest-link-suggestion",
            &mut norm,
        );
        assert_eq!(path, "/api/tracking/session/<UUID_1>/quest-link-suggestion");
    }

    #[test]
    fn numeric_duration_projects_to_sentinel() {
        let body = json!({"sessions": [{"duration": 42, "id": 1}], "duration": 7.0});
        let projected = project_session_duration(body);
        assert_eq!(projected["duration"], json!("<SESSION_DURATION>"));
        assert_eq!(
            projected["sessions"][0]["duration"],
            json!("<SESSION_DURATION>")
        );
        assert_eq!(projected["sessions"][0]["id"], json!(1));
    }

    #[test]
    fn empty_body_is_null_and_binary_projects_byte_length() {
        let mut norm = Normalizer::new();
        assert_eq!(
            normalise_body(b"", Some("application/json"), &mut norm),
            Value::Null
        );
        assert_eq!(
            normalise_body(&[0u8, 1, 2], Some("image/png"), &mut norm),
            json!({"_binary": true, "byte_length": 3})
        );
    }
}
